panc/src/main/java/org/quattor/pan/dml/data/ProtectedListResource.java

```diff
@@ -39,13 +43,28 @@ public Element put(int index, Element newValue) {
 	}
 
 	@Override
+	public void append(Element e) {
+		throw CompilerError.create(MSG_ILLEGAL_WRITE_TO_PROTECTED_LIST);
+	}
+
+	@Override
+	public void prepend(Element e) {
+		throw CompilerError.create(MSG_ILLEGAL_WRITE_TO_PROTECTED_LIST);
+	}
+
+	@Override
 	public int size() {
 		return baseList.size();
 	}
 
 	@Override
 	public Resource.Iterator iterator() {
-		return baseList.iterator();
+		return baseList.protectedIterator();
```

Comment:
No, that is exactly one of the issues I have debugged: .protect() wraps only the top-level reference, it does not alter the data behind that reference. So when you're simply iterating over a protected resource, the iteration blows away the protection.

I've also thought about changing the data, by embedding the "I am protected" bit of information into HashResource/ListResource, but that may require deeper changes to other parts of the compiler.